File: fore-core/src/main/java/co/early/fore/core/logging/SystemLogger.java
package co.early.fore.core.logging;


import co.early.fore.core.utils.TextPaddingUtils;

public class SystemLogger implements Logger {

    private int longestTagLength = 0;

    @Override
    public void e(String tag, String message) {
        System.out.println("(E) " + padTagWithSpace(tag) + "|" + message);
    }

    @Override
    public void w(String tag, String message) {
        System.out.println("(W) " + padTagWithSpace(tag) + "|" + message);
    }

    @Override
    public void i(String tag, String message) {
        System.out.println("(I) " + padTagWithSpace(tag) + "|" + message);
    }

    @Override
    public void d(String tag, String message) {
        System.out.println("(D) " + padTagWithSpace(tag) + "|" + message);
    }

    @Override
    public void v(String tag, String message) {
        System.out.println("(V) " + padTagWithSpace(tag) + "|" + message);
    }


    @Override
    public void e(String tag, String message, Throwable throwable) {
        e(tag, message);
        System.out.println(throwable);
    }

    @Override
    public void w(String tag, String message, Throwable throwable) {
        w(tag, message);
        System.out.println(throwable);
    }

    @Override
    public void i(String tag, String message, Throwable throwable) {
        i(tag, message);
        System.out.println(throwable);
    }

    @Override
    public void d(String tag, String message, Throwable throwable) {
        d(tag, message);
        System.out.println(throwable);
    }

    @Override
    public void v(String tag, String message, Throwable throwable) {
        v(tag, message);
        System.out.println(throwable);
    }


    private String padTagWithSpace(String tag){

        longestTagLength = Math.max(longestTagLength, tag.length() + 1);

        if (longestTagLength != tag.length()) {
            return TextPaddingUtils.padText(tag, longestTagLength, TextPaddingUtils.Padding.END, ' ');
        }else{
            return tag;
        }
    }
}

File: fore-core-kt/src/main/java/co/early/fore/kt/core/logging/SystemLogger.kt
package co.early.fore.kt.core.logging

class SystemLogger :
        Logger,
        TagInferer by TagInfererImpl() {

    private var longestTagLength = 0

    override fun e(message: String) {
        e(inferTag(), message)
    }

    override fun w(message: String) {
        w(inferTag(), message)
    }

    override fun i(message: String) {
        i(inferTag(), message)
    }

    override fun d(message: String) {
        d(inferTag(), message)
    }

    override fun v(message: String) {
        v(inferTag(), message)
    }

    override fun e(message: String, throwable: Throwable) {
        e(inferTag(), message, throwable)
    }

    override fun w(message: String, throwable: Throwable) {
        w(inferTag(), message, throwable)
    }

    override fun i(message: String, throwable: Throwable) {
        i(inferTag(), message, throwable)
    }

    override fun d(message: String, throwable: Throwable) {
        d(inferTag(), message, throwable)
    }

    override fun v(message: String, throwable: Throwable) {
        v(inferTag(), message, throwable)
    }

    override fun e(tag: String, message: String) {
        println("(E) " + padTagWithSpace(tag) + "|" + message)
    }

    override fun w(tag: String, message: String) {
        println("(W) " + padTagWithSpace(tag) + "|" + message)
    }

    override fun i(tag: String, message: String) {
        println("(I) " + padTagWithSpace(tag) + "|" + message)
    }

    override fun d(tag: String, message: String) {
        println("(D) " + padTagWithSpace(tag) + "|" + message)
    }

    override fun v(tag: String, message: String) {
        println("(V) " + padTagWithSpace(tag) + "|" + message)
    }

    override fun e(tag: String, message: String, throwable: Throwable) {
        e(tag, message)
        println(throwable)
    }

    override fun w(tag: String, message: String, throwable: Throwable) {
        w(tag, message)
        println(throwable)
    }

    override fun i(tag: String, message: String, throwable: Throwable) {
        i(tag, message)
        println(throwable)
    }

    override fun d(tag: String, message: String, throwable: Throwable) {
        d(tag, message)
        println(throwable)
    }

    override fun v(tag: String, message: String, throwable: Throwable) {
        v(tag, message)
        println(throwable)
    }

    private fun padTagWithSpace(tag: String): String? {
        longestTagLength = Math.max(longestTagLength, tag.length + 1)
        return if (longestTagLength != tag.length) {
            co.early.fore.core.utils.text.TextPadder.padText(tag, longestTagLength, co.early.fore.core.utils.text.TextPadder.Pad.RIGHT, ' ')
        } else {
            tag
        }
    }
}
